llarp/router/rc_lookup_handler.cpp

```diff
@@ -0,0 +1,323 @@
+      }
+      shouldDoLookup = itr_pair.second;
+    }
+
+    if (shouldDoLookup)
+    {
+
+      auto fn = std::bind(&RCLookupHandler::HandleDHTLookupResult,
+                            this,
+                            router,
+                            std::placeholders::_1);
+
+      // if we are a client try using the hidden service endpoints
+      if(!isServiceNode)
+      {
+        bool sent = false;
+        LogInfo("Lookup ", router, " anonymously");
+        _hiddenServiceContext->ForEachService(
+            [&](const std::string &,
+                const std::shared_ptr< service::Endpoint > &ep) -> bool {
+              const bool success = ep->LookupRouterAnon(router, fn);
+              sent               = sent || success;
+              return !success;
+            });
+        if(sent)
+          return;
+        LogWarn("cannot lookup ", router, " anonymously");
+      }
+
+      if (!_dht->impl->LookupRouter(router, fn))
+      {
+        FinalizeRequest(router, nullptr, RCRequestResult::RouterNotFound);
+      }
+    }
+  }
+
+  bool
+  RCLookupHandler::RemoteIsAllowed(const RouterID &remote) const
+  {
+    if(_strictConnectPubkeys.size() &&
+       _strictConnectPubkeys.count(remote) == 0 &&
+       !RemoteInBootstrap(remote))
+    {
+      return false;
+    }
+
+    util::Lock l(&_mutex);
+
+    if(useWhitelist
+       && whitelistRouters.find(remote) == whitelistRouters.end())
+    {
+      return false;
+    }
+
+    return true;
+  }
+
+  bool
+  RCLookupHandler::CheckRC(const RouterContact &rc) const
+  {
+    if (not RemoteIsAllowed(rc.pubkey))
+    {
+      return false;
+    }
+
+    if (not rc.Verify(_dht->impl->Now()))
+    {
+      return false;
+    }
+
+    // update dht if required
+    if(_dht->impl->Nodes()->HasNode(dht::Key_t{rc.pubkey}))
+    {
+      _dht->impl->Nodes()->PutNode(rc);
+    }
+
+    // update nodedb if required
+    if (rc.IsPublicRouter())
+    {
+      _nodedb->UpdateAsyncIfNewer(rc);
+    }
+
+    return true;
+  }
+
+  bool
+  RCLookupHandler::GetRandomWhitelistRouter(RouterID &router) const
+  {
+    util::Lock l(&_mutex);
+
+    const auto sz = whitelistRouters.size();
+    auto itr      = whitelistRouters.begin();
+    if(sz == 0)
+      return false;
+    if(sz > 1)
+      std::advance(itr, randint() % sz);
+    router = *itr;
+    return true;
+  }
+
+  bool
+  RCLookupHandler::CheckRenegotiateValid(RouterContact newrc, RouterContact oldrc)
+  {
+    // missmatch of identity ?
+    if(newrc.pubkey != oldrc.pubkey)
+      return false;
+
+    if(!RemoteIsAllowed(newrc.pubkey))
+      return false;
+
+
+    // TODO: check for other places that need updating the RC
+    return true;
+  }
+
+  void
+  RCLookupHandler::PeriodicUpdate(llarp_time_t now)
+  {
+    // try looking up stale routers
+    _nodedb->VisitInsertedBefore(
+        [&](const RouterContact &rc) {
+          if(HavePendingLookup(rc.pubkey))
+            return;
+          GetRC(rc.pubkey, nullptr);
```

Comment:
GetRC calls stuff that acquires a lock that is held while visiting.

Reply:
That's...gonna be fun to resolve.

Reply:
Addressed.